File: app/src/main/java/org/michaelbel/moviemade/data/injection/module/AppModule.java
package org.michaelbel.moviemade.data.injection.module;

import android.app.Application;
import android.content.Context;
import android.content.res.Resources;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {

    private Context context;

    public AppModule(Application app) {
        context = app;
    }

    @Provides
    @Singleton
    Context provideContext() {
        return context;
    }

    @Provides
    @Singleton
    Resources provideResources(Context context) {
        return context.getResources();
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/di/module/AppModule.kt
package org.michaelbel.moviemade.presentation.di.module

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import org.michaelbel.data.local.AppDatabase
import org.michaelbel.moviemade.core.local.SharedPrefs.NAME
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Provides
    @Singleton
    fun sharedPreferences(@ApplicationContext context: Context): SharedPreferences =
        context.getSharedPreferences(NAME, MODE_PRIVATE)

    @Provides
    @Singleton
    fun appDatabase(@ApplicationContext context: Context) = AppDatabase.getInstance(context)

    //region Dao

    @Provides
    fun trailersDao(db: AppDatabase) = db.trailersDao

    @Provides
    fun reviewsDao(db: AppDatabase) = db.reviewsDao

    @Provides
    fun keywordsDao(db: AppDatabase) = db.keywordsDao

    @Provides
    fun moviesDao(db: AppDatabase) = db.moviesDao

    @Provides
    fun usersDao(db: AppDatabase) = db.usersDao

    //endregion
}